### ISSUE TITLE
Cannot use a read-only upstream, since Spack tries to write on the filesystem

### ISSUE THREAD

**Issue by hartzell** [LLNL/spack#17723]
TL;DR: It seems that I **can not** chain to a read-only tree.

---

I build spack trees with a fairly set of packages; once built I consider them immutable.

One of my work situations involves AWS instances with a shared EFS filesystem that has horrible performance.  I'd like to build my trees in the EFS filesystem so that they're easily available but they'd take *forever*.

My current solution is to build them at their proper path but in isolation (e.g. in a Docker container), save the tree as a squashfs `.sqsh` file, copy the `.sqsh` file into the shared filesystem (one big file, performance is acceptable), and then use a loopback mount to make it available in its proper location (mounts managed via autofs).

It's a bunch of parts but they line up nicely end-to-end so the complexity isn't overwhelming.

In the end it works nicely and my "immutable" trees are truly read-only.  But....

It seems that I **can not** chain to a read-only tree.

If I:

- clone another copy of my spack repository;
- add an `etc/spack/upstreams.yaml` like so:

  ```yaml
  upstreams:
    spack-instance-1:
      install_tree: /blaheng/env/release/0.1.0
  ```

I end up with this:

```shellsession
$ . share/spack/setup-env.sh
$ spack find
==> Error: [Errno 30] Read-only file system: '/blaheng/env/release/0.1.0/.spack-db'
```

Disabling locks doesn't help:

```shellsession
$ spack --disable-locks find
==> Error: Refusing to disable locks: spack is group-writable.
Running a shared spack without locks is unsafe. You must restrict permissions on /blah/blort/tmp/spack-readonly-bug or enable locks.
$ chmod 755 .
$ spack --disable-locks find
==> Error: [Errno 30] Read-only file system: '/blaheng/env/release/0.1.0/.spack-db'
$
```

I ran the demo commands above with a slightly mismatched {up,down}stream pair of trees:

```shellsession
$ ./bin/spack --version
0.14.1-1934-cefb4ba01
$ /blaheng/env/release/0.1.0/bin/spack --version
0.14.2
```

I'm working on building a standard 'env' tree that up to date w.r.t. this tree and will revisit this.  I've seen the problem with a couple of different version pairings so I'm fairly confident that it's the mis-pairing is not the problem.

**Comment by smattis-nnl** [comment]
I am also having this same issue with the current develop version of spack. I have a read-only upstream, and when I try to build anything I get  an error about  not having write permissions for the `.spack-db' file in the upstream spack distribution.

**Comment by wspear** [comment]
I'm having this issue as well.
Pointing my spack install to a read only /spack/opt . After I set up my upstreams.yaml I get this:
```
spack find
==> Error: [Errno 30] Read-only file system: '/spack/opt/.spack-db'
```

**Comment by becker33** [comment]
Can someone run with `spack -d` to get a stacktrace of the failure? @wspear ?

**Comment by wspear** [comment]
@becker33 behold!
```
spack -d find
==> [2022-03-03-12:05:08.364109] Imported find from built-in commands
==> [2022-03-03-12:05:08.365994] Imported find from built-in commands
==> [2022-03-03-12:05:08.367400] Reading config file /global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/etc/spack/defaults/config.yaml
==> [2022-03-03-12:05:08.391756] Reading config file /global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/etc/spack/defaults/bootstrap.yaml
==> [2022-03-03-12:05:08.401581] Reading config file /global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/etc/spack/linux/upstreams.yaml
Traceback (most recent call last):
  File "/global/homes/w/wspear/PERLMUTTER/spack-daisychain/spack/bin/spack", line 98, in <module>
    sys.exit(spack.main.main())
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/main.py", line 916, in main
    return _main(argv)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/main.py", line 871, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/main.py", line 899, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/main.py", line 556, in _invoke_command
    return_val = command(parser, args)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/cmd/find.py", line 212, in find
    _find(parser, args)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/cmd/find.py", line 217, in _find
    results = args.specs(**q_args)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/cmd/common/arguments.py", line 78, in _specs
    return spack.store.db.query(**kwargs)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/llnl/util/lang.py", line 829, in __getattr__
    return getattr(self.ref_function(), name)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/store.py", line 225, in _store_db
    return store.db
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/llnl/util/lang.py", line 799, in __getattr__
    return getattr(self.instance, name)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/llnl/util/lang.py", line 789, in instance
    self._instance = self.factory()
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/store.py", line 206, in _store
    return Store(root=root,
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/store.py", line 159, in __init__
    root, upstream_dbs=retrieve_upstream_dbs())
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/store.py", line 271, in retrieve_upstream_dbs
    return _construct_upstream_dbs_from_install_roots(install_roots)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/store.py", line 279, in _construct_upstream_dbs_from_install_roots
    next_db = spack.database.Database(
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/database.py", line 354, in __init__
    fs.mkdirp(self._db_dir)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/llnl/util/filesystem.py", line 627, in mkdirp
    raise e
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/llnl/util/filesystem.py", line 590, in mkdirp
    os.makedirs(path)
  File "/usr/lib/python3.8/os.py", line 223, in makedirs
    mkdir(name, mode)
OSError: [Errno 30] Read-only file system: '/spack/opt/.spack-db'
```

**Comment by eugeneswalker** [comment]
```
$> cat ~/.spack/upstreams.yaml
upstreams:
  spack-instance-1:
    install_tree: /spack
```

```
$> spack -d find
==> [2022-03-03-12:06:06.976185] Imported find from built-in commands
==> [2022-03-03-12:06:06.977039] Imported find from built-in commands
==> [2022-03-03-12:06:06.978153] Reading config file /home/users/ewalker/spack-test/etc/spack/defaults/config.yaml
==> [2022-03-03-12:06:06.990835] Reading config file /home/users/ewalker/spack-test/etc/spack/defaults/bootstrap.yaml
==> [2022-03-03-12:06:06.995989] Reading config file /home/users/ewalker/.spack/upstreams.yaml
Traceback (most recent call last):
  File "/home/users/ewalker/spack-test/bin/spack", line 98, in <module>
    sys.exit(spack.main.main())
  File "/home/users/ewalker/spack-test/lib/spack/spack/main.py", line 916, in main
    return _main(argv)
  File "/home/users/ewalker/spack-test/lib/spack/spack/main.py", line 871, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/home/users/ewalker/spack-test/lib/spack/spack/main.py", line 899, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/home/users/ewalker/spack-test/lib/spack/spack/main.py", line 556, in _invoke_command
    return_val = command(parser, args)
  File "/home/users/ewalker/spack-test/lib/spack/spack/cmd/find.py", line 212, in find
    _find(parser, args)
  File "/home/users/ewalker/spack-test/lib/spack/spack/cmd/find.py", line 217, in _find
    results = args.specs(**q_args)
  File "/home/users/ewalker/spack-test/lib/spack/spack/cmd/common/arguments.py", line 78, in _specs
    return spack.store.db.query(**kwargs)
  File "/home/users/ewalker/spack-test/lib/spack/llnl/util/lang.py", line 829, in __getattr__
    return getattr(self.ref_function(), name)
  File "/home/users/ewalker/spack-test/lib/spack/spack/store.py", line 225, in _store_db
    return store.db
  File "/home/users/ewalker/spack-test/lib/spack/llnl/util/lang.py", line 799, in __getattr__
    return getattr(self.instance, name)
  File "/home/users/ewalker/spack-test/lib/spack/llnl/util/lang.py", line 789, in instance
    self._instance = self.factory()
  File "/home/users/ewalker/spack-test/lib/spack/spack/store.py", line 206, in _store
    return Store(root=root,
  File "/home/users/ewalker/spack-test/lib/spack/spack/store.py", line 159, in __init__
    root, upstream_dbs=retrieve_upstream_dbs())
  File "/home/users/ewalker/spack-test/lib/spack/spack/store.py", line 271, in retrieve_upstream_dbs
    return _construct_upstream_dbs_from_install_roots(install_roots)
  File "/home/users/ewalker/spack-test/lib/spack/spack/store.py", line 279, in _construct_upstream_dbs_from_install_roots
    next_db = spack.database.Database(
  File "/home/users/ewalker/spack-test/lib/spack/spack/database.py", line 354, in __init__
    fs.mkdirp(self._db_dir)
  File "/home/users/ewalker/spack-test/lib/spack/llnl/util/filesystem.py", line 627, in mkdirp
    raise e
  File "/home/users/ewalker/spack-test/lib/spack/llnl/util/filesystem.py", line 590, in mkdirp
    os.makedirs(path)
  File "/usr/local/miniconda3/lib/python3.9/os.py", line 225, in makedirs
    mkdir(name, mode)
OSError: [Errno 30] Read-only file system: '/spack/.spack-db'

**Comment by becker33** [comment]
Where is the `.spack-db` directory associated with the upstream?

**Comment by becker33** [comment]
Or what is the `install_tree:root` value for the upstream?

**Comment by wspear** [comment]
For mine it's in:  /spack/opt/spack/.spack-db/

**Comment by becker33** [comment]
Change your `upstreams` config to 

```
upstreams:
  spack-instance-1:
    install_tree: /spack/opt/spack
```

**Comment by becker33** [comment]
After that, does it still fail?

**Comment by wspear** [comment]
Same issue. And the path that's failing doesn't seem to have changed despite updating the path in upstreams.yaml

```
wspear@login20:/global/u2/w/wspear/PERLMUTTER/spack-daisychain$ spack -d find
==> [2022-03-03-13:16:28.876576] Imported find from built-in commands
==> [2022-03-03-13:16:28.878398] Imported find from built-in commands
==> [2022-03-03-13:16:28.879828] Reading config file /global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/etc/spack/defaults/config.yaml
==> [2022-03-03-13:16:28.904313] Reading config file /global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/etc/spack/defaults/bootstrap.yaml
==> [2022-03-03-13:16:28.914081] Reading config file /global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/etc/spack/linux/upstreams.yaml
Traceback (most recent call last):
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/bin/spack", line 98, in <module>
    sys.exit(spack.main.main())
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/main.py", line 916, in main
    return _main(argv)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/main.py", line 871, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/main.py", line 899, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/main.py", line 556, in _invoke_command
    return_val = command(parser, args)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/cmd/find.py", line 212, in find
    _find(parser, args)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/cmd/find.py", line 217, in _find
    results = args.specs(**q_args)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/cmd/common/arguments.py", line 78, in _specs
    return spack.store.db.query(**kwargs)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/llnl/util/lang.py", line 829, in __getattr__
    return getattr(self.ref_function(), name)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/store.py", line 225, in _store_db
    return store.db
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/llnl/util/lang.py", line 799, in __getattr__
    return getattr(self.instance, name)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/llnl/util/lang.py", line 789, in instance
    self._instance = self.factory()
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/store.py", line 206, in _store
    return Store(root=root,
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/store.py", line 159, in __init__
    root, upstream_dbs=retrieve_upstream_dbs())
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/store.py", line 271, in retrieve_upstream_dbs
    return _construct_upstream_dbs_from_install_roots(install_roots)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/store.py", line 279, in _construct_upstream_dbs_from_install_roots
    next_db = spack.database.Database(
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/spack/database.py", line 354, in __init__
    fs.mkdirp(self._db_dir)
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/llnl/util/filesystem.py", line 627, in mkdirp
    raise e
  File "/global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/lib/spack/llnl/util/filesystem.py", line 590, in mkdirp
    os.makedirs(path)
  File "/usr/lib/python3.8/os.py", line 223, in makedirs
    mkdir(name, mode)
OSError: [Errno 30] Read-only file system: '/spack/opt/.spack-db'
wspear@login20:/global/u2/w/wspear/PERLMUTTER/spack-daisychain$ cat ~/.spack/upstreams.yaml
upstreams:
  spack-instance-1:
    install_tree: /spack/opt/spack

```

**Comment by wspear** [comment]
It was taking its config from /global/u2/w/wspear/PERLMUTTER/spack-daisychain/spack/etc/spack/linux/upstreams.yaml instead of .spack/upstreams.yaml. Is that how scoping is supposed to work?  When I edited the one in etc `spack find` worked.

**Comment by becker33** [comment]
@eugeneswalker does changing the path resolve the issue for you as well?

**Comment by tgamblin** [comment]
I think regardless of whether this fixes things for @eugeneswalker and @wspear we should make it so that the DB doesn't attempt to create anything if it finds itself on a read-only filesystem.

We should probably fail with some intuitive error if we don't find `.spack-db` for an upstream or read-only filesystem, rather than trying to create it and failing.

**Comment by bollig** [comment]
Raised this in slack and thought it would help others that find their way here: https://gist.github.com/bollig/aab44a2e5774d80d992c2974f8b2ea7b

I use this to root deploy spack where root has control over upstream packages and users have read-only access. Users get read/write control over packages that install to their own $HOME/.spack.

**Comment by wspear** [comment]
@tgamblin beyond that, maybe there should be a useful error message or at least a warning if a .spack-db is not found regardless of the writability of the filesystem. I'd think in most cases users intend to point their upstreams to a valid install_tree:root.
